controlplane/kubeadm/controllers/helpers.go

```diff
@@ -246,3 +248,87 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 	}
 	return nil
 }
+
+// MachinesNeedingRollout return a list of machines that need to be upgraded.
+func (r *KubeadmControlPlaneReconciler) MachinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
+	machinesToUpgrade := internal.NewFilterableMachineCollection()
+
+	now := metav1.Now()
+	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
+		oldMachines := c.Machines.Filter(machinefilters.OlderThan(c.KCP.Spec.UpgradeAfter))
+		// Ignore deleted machines.
+		// TODO: This should be removed once MHC remediation is merged.
+		oldMachines = oldMachines.Filter(func(machine *clusterv1.Machine) bool {
+			return machine.DeletionTimestamp.IsZero()
+		})
+		for _, m := range oldMachines {
+			machinesToUpgrade.Insert(m)
+		}
+	}
+
+	for _, machine := range c.Machines {
+		// TODO: This should be removed once MHC remediation is merged.
+		if !machine.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		if *machine.Spec.Version != c.KCP.Spec.Version {
+			machinesToUpgrade.Insert(machine)
+			continue
+		}
+
+		ref := machine.Spec.Bootstrap.ConfigRef
+		if ref == nil {
+			continue
+		}
+
+		machineCfg := &bootstrapv1.KubeadmConfig{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: c.KCP.Namespace}, machineCfg); err != nil {
+			continue
+		}
+
+		kcpSpec := createKcpKubeadmConfig(machineCfg.Spec, c.KCP.Spec.KubeadmConfigSpec)
+
+		if !reflect.DeepEqual(machineCfg.Spec, kcpSpec) {
+			machinesToUpgrade.Insert(machine)
+			continue
+		}
+
+		infraObj, err := external.Get(ctx, r.Client, &machine.Spec.InfrastructureRef, machine.Spec.InfrastructureRef.Namespace)
+		if err != nil {
+			//
+			continue
+		}
+
+		clonedFromName, existName := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
+		clonedFromGroupKind, existGroupKind := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !existName || !existGroupKind {
+			// All KCP cloned infra machines should have this annotation.
+			// Missing the annotation here may be due to older version machines or adopted machines.
+			continue
+		}
+
+		if clonedFromName != c.KCP.Spec.InfrastructureTemplate.Name || clonedFromGroupKind != c.KCP.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
```

Comment:
Maybe let's have the second if condition on a new line for readability? Let's also add a comment explaining why we need to check these this way